### ISSUE TITLE
[@types/mongoose] `this` has type `any` in a `pre` hook

### ISSUE THREAD

**Issue by osenvosem** [DefinitelyTyped/DefinitelyTyped#24656]
Can you please show me the way one is supposed to type `this` in a `.pre` hook?

```
[ts] 'this' implicitly has type 'any' because it does not have a type annotation.
```

![screen shot 2018-03-31 at 15 24 51](https://user-images.githubusercontent.com/6762673/38163156-de60b1b2-34f7-11e8-8476-04e83b042fa6.png)

We need a way to specify the value of `this` in a hook.

We can change the types this way:

```ts
pre<T>(method: string, parallel: boolean, fn: HookAsyncCallback<T>, errorCb?: HookErrorCallback): this;
pre<T>(method: string, fn: HookSyncCallback<T>, errorCb?: HookErrorCallback): this;
//...
interface HookSyncCallback<T> {
  (this: T, next: HookNextFunction): any;
}
interface HookAsyncCallback<T> {
  (this: T, next: HookNextFunction, done: HookDoneFunction): any;
}
```

And use it this way:

```ts
userSchema.pre<TUserDocument>("save", function(next) {
  //...
})
```

@horiuchi, @sindrenm, @lukasz-zak, @Alorel, @jendrikw, @ethanresnick, @vologab

**Comment by Alorel** [comment]
Note: I'm hardly an expert on Typescript generics.

It should probably be `pre<T = any>(...)` then because I *think* forcing the function to have a type is a breaking change. This itself would be a breaking `@types` change because I tried using the syntax in one of my PRs and, since the current type definitions have the typescript version set to 2.3, the CI failed on old TS versions.

**Comment by osenvosem** [comment]
@Alorel good point. Maybe others also have some ideas on this one.

**Comment by ethanresnick** [comment]
Also, in some `pre` hooks, I believe `T` can be a Query object, rather than a document (e.g., `pre("findOne", ....)`. So, I think awesome types would take the middleware name into account and constrain/default `T` appropriately. This would be hugely useful for catching errors where the middleware function accidentally assumes T will be a Document. Something like:

```
// Document middleware signature with type parameter constrained to, and defaulting to, Document
pre<T extends Document = Document>(method: "init" | "validate" | "save" | "remove", // rest of args here): this

// Other doc overloads...

// Query middleware signature
pre<T extends Query<any> = Query<any>>(method: "count" | "findOne" | "findOneAndRemove" | "findOneAndUpdate" | "update", // rest of args here): this

// Other query overloads...

// maybe a fallback signature for the string case when the method isn't a literal
pre<T extends Document | Query<any> = Document | Query<any>>(method: string, // rest of args here): this
```

**Comment by ethanresnick** [comment]
@osenvosem Yup, you're right. I'd add a signature for those too! There are gonna end up being a ton of signatures, but that's fine.

**Comment by osenvosem** [comment]
So this is the way it may look like. The only thing that remains unclear is what the "Aggregation object" is. It mentioned in the documentation, but there is no type definition for it and I can't find anything in the docs about what it is. 🤔

```ts
    pre<T extends Document = Document>(
      method: "init" | "validate" | "save" | "remove",
      parallel: boolean,
      fn: HookAsyncCallback<T>,
      errorCb?: HookErrorCallback
    ): this;
    pre<T extends Query<any> = Query<any>>(
      method:
        | "count"
        | "find"
        | "findOne"
        | "findOneAndRemove"
        | "findOneAndUpdate"
        | "update",
      parallel: boolean,
      fn: HookAsyncCallback<T>,
      errorCb?: HookErrorCallback
    ): this;
    pre<T = any>(
      method: "aggregate",
      parallel: boolean,
      fn: HookAsyncCallback<T>,
      errorCb?: HookErrorCallback
    ): this;
    pre<T extends Model<Document> = Model<Document>>(
      method: "insertMany",
      parallel: boolean,
      fn: HookAsyncCallback<T>,
      errorCb?: HookErrorCallback
    ): this;
    pre<T extends Document | Model<Document> | Query<any> | any>(
      method: string,
      parallel: boolean,
      fn: HookAsyncCallback<T>,
      errorCb?: HookErrorCallback
    ): this;

    pre<T extends Document = Document>(
      method: "init" | "validate" | "save" | "remove",
      fn: HookSyncCallback<T>,
      errorCb?: HookErrorCallback
    ): this;
    pre<T extends Query<any> = Query<any>>(
      method:
        | "count"
        | "find"
        | "findOne"
        | "findOneAndRemove"
        | "findOneAndUpdate"
        | "update",
      fn: HookSyncCallback<T>,
      errorCb?: HookErrorCallback
    ): this;
    pre<T = any>(
      method: "aggregate",
      fn: HookSyncCallback<T>,
      errorCb?: HookErrorCallback
    ): this;
    pre<T extends Model<Document> = Model<Document>>(
      method: "insertMany",
      fn: HookSyncCallback<T>,
      errorCb?: HookErrorCallback
    ): this;
    pre<T extends Document | Model<Document> | Query<any> | any>(
      method: string,
      fn: HookSyncCallback<T>,
      errorCb?: HookErrorCallback
    ): this;
```

**Comment by ethanresnick** [comment]
@osenvosem Yup, that looks great to me! If you test it and it works, I would be +1 on a PR with those changes

**Comment by wooandoo** [comment]
we can't call 

```js
schema.pre('remove', { query: true, document: true }, fn)
```

**the error** is: Argument of type '{ query: boolean; }' is not assignable to parameter of type 'boolean'

**Comment by gpaglia** [comment]
I think I have the very same problem on virtual fields, with typescript 3.1.6, @types/mongoose 5.3.0 and mongoose  5.3.9.
Sorry if I overloked something, but I'm pretty new to mongoose on typescript...

I'm trying to define a virtual field on a subdocument and I get the complaint:

> [ts] 'this' implicitly has type 'any' because it does not have a type annotation.

This is the snippet:
`ClientSubscriptionSchema.virtual('clientId').get(function () {
    const me = this;
});
`
Moreover, as I need to refer to the parent() of the subdocument, I have found no way to access the parent document as the typescript types do not expose the EmbededDocument type with its parent() method.

In fact, to bypass the 'this' problem, I have tried to make use of an explicit this parameter in the getter definition, but the Document interface does not expose the parent() method.

`export interface IClientSubscriptionModel extends IClientSubscription, Document {
}
ClientSubscriptionSchema.virtual('clientId').get(function (this: IClientSubscriptionModel) {
    const me = this.parent();
});
`
Any hint?

thanks a lot
gianni
